gnmi_subscribe/gnmi_subscribe.go

```diff
@@ -77,97 +79,160 @@ func main() {
+
+	subscriptions, err := assembleSubscriptions(*streamOnChange, *sampleInterval, pbPathList)
 	if err != nil {
 		log.Exitf("Error assembling subscriptions: %v", err)
 	}
 
 	request := &pb.SubscribeRequest{
 		Request: &pb.SubscribeRequest_Subscribe{
 			Subscribe: &pb.SubscriptionList{
-				Encoding:     pb.Encoding(encoding),
+				Encoding:     encoding,
 				Mode:         subscriptionListMode,
 				Subscription: subscriptions,
+				UpdatesOnly:  *updatesOnly,
 			},
 		},
 	}
 	utils.PrintProto(request)
 	if err := subscribeClient.Send(request); err != nil {
 		log.Exitf("Failed to send request: %v", err)
 	}
+
 	switch subscriptionListMode {
 	case pb.SubscriptionList_STREAM:
-		stream()
+		if err := stream(subscribeClient); err != nil {
+			log.Exitf("Error using STREAM mode: %v", err)
+		}
 	case pb.SubscriptionList_POLL:
-		poll()
+		if err := poll(subscribeClient, *updatesOnly, pollUser); err != nil {
+			log.Exitf("Error using POLL mode: %v", err)
+		}
 	case pb.SubscriptionList_ONCE:
-		once()
+		if err := once(subscribeClient); err != nil {
+			log.Exitf("Error using ONCE mode: %v", err)
+		}
 	}
 }
 
-func stream() {
+func pollUser() {
+	fmt.Scanln()
+}
 
+func stream(subscribeClient gnmi.GNMI_SubscribeClient) error {
+	for {
+		res, err := subscribeClient.Recv()
+		if err == io.EOF {
+			log.Info("Stream closed by target")
+			return nil
+		}
+		if err != nil {
+			return err
+		}
+		switch res.Response.(type) {
+		case *pb.SubscribeResponse_SyncResponse:
+			log.Info("SyncResponse received")
+		case *pb.SubscribeResponse_Update:
+			utils.PrintProto(res)
+		default:
+			return errors.New("Unexpected response type")
+		}
+	}
 }
 
-func poll() {
+func poll(subscribeClient gnmi.GNMI_SubscribeClient, updatesOnly bool, pollInput func()) error {
+	ready := make(chan bool, 1)
+	ready <- true
+	pollRequest := &pb.SubscribeRequest{Request: &pb.SubscribeRequest_Poll{}}
+	if updatesOnly {
+		res, err := subscribeClient.Recv()
+		if err != nil {
+			return err
+		}
+		if syncRes := res.GetSyncResponse(); !syncRes {
+			return errors.New("-updates_only flag is set but failed to receive SyncResponse first for POLL mode")
+		}
+		log.Info("SyncResponse received")
+	}
+	for {
+		select {
+		case <-ready:
+			log.Info("Press enter to poll")
+			pollInput()
+			if err := subscribeClient.Send(pollRequest); err != nil {
+				return err
+			}
+			utils.LogProto(pollRequest)
+		default:
+			res, err := subscribeClient.Recv()
+			if err == io.EOF {
+				return nil
+			}
+			if err != nil {
+				return err
+			}
+			switch res.Response.(type) {
+			case *pb.SubscribeResponse_SyncResponse:
+				log.Info("SyncResponse received")
+				ready <- true
+			case *pb.SubscribeResponse_Update:
+				utils.PrintProto(res)
+			default:
+				return errors.New("Unexpected response type")
+			}
+		}
+	}
 
 }
 
-func once() {
-	
+func once(subscribeClient gnmi.GNMI_SubscribeClient) error {
+	for {
+		res, err := subscribeClient.Recv()
+		if err != nil {
```

Comment:
What if you get EOF? You checked for it in the poll function earlier.

Reply:
Also, all of this seems common to poll also. Could break out common code into a function to be called by both.

Reply:
My mistake, I should have added it, agree on common code, will refactor it